### ISSUE TITLE
Look into what happens when you flag an entire Kiwi Game for garbage collection

### ISSUE THREAD

**Issue by BenjaminHarding** [gamelab/kiwi.js#18]
Look into what happens when you flag an entire Kiwi Game for garbage collection and then fix/make changes accordingly.

**Comment by jdrorrer** [comment]
+1!

I have a question that I'm not totally sure if it's related, but I have a feeling it's tied to garbage collection... I have the Zoe Dress-up game included in a single-page app and it works when I initially navigate to the template/route it's on, but if I navigate to another page and then go back it does not load and this error comes up: 

`Uncaught TypeError: Cannot read property 'isAudio' of undefined` pops up right after `Kiwi.AudioLibrary: Rebuilding Audio Library` fires from the debugger.

I tried clearing/destroying all the game states when the template is destroyed as well as resetting the fileStore to undefined or an empty object, but neither has done the trick.  Is this a familiar problem?  Any ideas what it could be?

**Comment by BenjaminHarding** [comment]
I have not encountered that error before, or if so I at least cannot remember it. 

Which version of Kiwi are you using? `Kiwi.VERSION`

Have a go at using the latest 1.3.0, since I believe we have upgraded some areas around loading files and the fileStore in 1.3.0.

**Comment by jdrorrer** [comment]
@BenjaminHarding thanks for the quick response!

I was using version 1.1.1, but I upgraded to 1.3.0 and get the same error.  However, I also have kiwipreloader.js and it continues to try loading other files even after the isAudio error, and it produces this new error: `Uncaught TypeError: Cannot set property '_onCompleteCalled' of undefined`

One other thing I found is another dev I'm working with commented out this line of code:
```
Keyboard.prototype.reset = function () {
                for (var index in this._keys) {
                    // this._keys[index].reset();
                }
            };
            return Keyboard;
```
Not sure if that line could possibly be related, or even what it does.  However, when I uncomment this line, the Zoe DressUp game does not load at all and I get the following error: `Uncaught TypeError: this._keys[index].reset is not a function`

**Comment by BenjaminDRichards** [comment]
OK, I've done a deep delve into the bowels of KiwiJS and come to some conclusions.

### Recreation

This issue is easy enough to recreate. I just copied the `TemplateGame` folder from the KiwiJS repo and copied the applicable code from `src/game.js` to the end of `src/states/play.js: TemplateGame.Play.create()`. Sure enough, when a new game tries to boot with old states, it gives the exact same errors regarding audio and the `_onCompleteCalled` tween issue.

### Causes

Essentially, the preloader state is remembering the old game; but the old game has forgotten the preloader state, so the two try to exchange data and get confused.

It's possible to fix this by making the preloader state a class, not an instance; so instead of adding `TemplateGame.Loader` to the game, you'd add `new TemplateGame.Loader()` or some such. This means you get a new state every time. I've tested this, and it works.

However, there is a problem with this. The old game doesn't relinquish a number of contexts, including audio and WebGL and possibly others. Browsers only supply a limited number of these contexts to any given page. So after 6 game reboots, the game will hit another, harder error and stop again.

### Solutions

The simplest fix is just to create an iframe for each game. This is what we do behind the scenes at [www.kiwijs.org](www.kiwijs.org). When the game must be recreated, simply flush the old iframe entirely, and replace it with a new one. In this way, the game document context never has more than one call to the necessary context types.

If you want to simply stop the game from updating and leave it in the background for a while, it is theoretically possible to deactivate its `requestAnimationFrame` or RAF. There is a bug in the current release version of KiwiJS that prevents this from working, but it has been fixed in recent v1.3.1 commits. We intend to make this a part of the v1.4.0 release. The code in question is simply `game.raf.stop()` and `game.raf.start()`.

The final solution is the Herculean approach: we fight the hydra. KiwiJS was not originally architected for multiple games in the same context. It has a lot of connections to the window via event listeners and contexts. As we've never tried to run multiple games in parallel or in series within the same context, we've never encountered these issues before.

Some shutdown elements are already present; for example, there are `stop()` methods on the RAF, and on keyboard, mouse, and touch input. However, the current implementations don't work as desired. The RAF has already been fixed. The input listeners need to be changed from anonymous functions to fixed functions so they can be properly removed. There is also a visibility change event implementing by the Stage. We also need to ensure that contexts are released upon shutdown, if that is even possible. A proper shutdown method must ensure that all these function as intended.

The hydra fight has been postponed while we sharpen our weapons.

### Conclusion

If you want to use multiple KiwiJS games, iframes are your best bet. Other approaches are outside our current operational scope.

**Comment by jdrorrer** [comment]
@BenjaminDRichards You, sir, are the man!  Thank you so much for your detailed response... it makes perfect sense now and works flawlessly!  I owe you a beer or twelve... :)

In case this helps someone with the same issue in the future, the next issue I ran into (unrelated to kiwi.js) is dynamically setting the height of the iframe's parent container.  This is the best answer I found to that issue: http://www.h3xed.com/web-development/how-to-make-a-responsive-100-width-youtube-iframe-embed... the key is setting the padding-bottom to a percentage equal to your game's aspect ratio and setting the parent container's height to 0, so for example my game's height is set to 1024 and width is 768, so I set my padding-bottom equal to 133.33333% or 1024/768.
